### PR TITLE
Test scenario checking EK on TPM

### PR DESCRIPTION
**Pull request by Koncpa**
Scenarios verify genuine of TPM via validating EK cert.
First of these new scenarios is related with verify
EK cert against CA certs. This test used own generated endorsement key. Second scenario verify genuine of TPM by custom script.

Fix issue #105

### CODE REVIEW

File: functional/ek-cert-use-ek_check_script/main.fmf [hunk 1]
```diff
@@ -0,0 +1,28 @@
+summary:  Test of keylime checking of endorsment key by script
+description: |
+    Running all services on localhost.
+    Uses certificates generated by keylime.
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/main.fmf:4]: Is this related to TLS certificates or EK certificate? Please clarify.

Comment (kkaarreell) [reply functional/ek-cert-use-ek_check_script/main.fmf:4]: Also add that 'Configures agent to validate EK cert through ek_check_script

File: functional/ek-cert-use-ek_check_script/main.fmf [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    Starts verifier, registrar, agent.
+    Add script for verify EK.
+    Add keylime agent and validating EK by script.
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/main.fmf:7]: Add keylime agent and test EK cert has been validated by the configured script.

File: functional/ek-cert-use-ek_check_script/test.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        rlRun "limeWaitForVerifier"
+        rlRun "limeStartRegistrar"
+        rlRun "limeWaitForRegistrar"
+        rlRun "limeStartAgent"
+        rlRun "limeWaitForAgentRegistration ${AGENT_ID}"
+        # create allowlist and excludelist
+        limeCreateTestLists
+    rlPhaseEnd
+
+    # its possible, thats need to be need in different test scenario
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/test.sh:41]: This comment, what is it related to?

File: functional/ek-cert-use-ek_check_script/test.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    rlPhaseStartTest "Add keylime agent and check genuine of TPM via ek_check_script option"
+        rlRun "cat > /var/lib/keylime/check_ek_script.sh <<_EOF
+#!/bin/sh
+echo AGENT_UUID=$AGENT_UUID
+echo ENDORSEMENT_KEY=$EK
+echo ENDORSEMENT_KEY_TPM=$EK_TPM
+echo ENDORSEMENT_KEY_CERTIFICATES=$EK_CERT
+echo PROVKEYS=$PROVKEYS
+_EOF" 
+        rlRun "chmod 755 /var/lib/keylime/check_ek_script.sh"
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/test.sh:51]: I think better to set this script to be owned by `keylime.keylime` and have permissions `500`.

File: functional/ek-cert-use-ek_check_script/test.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        #veryfing of ek cert via own custom script
+        rlRun "limeUpdateConf tenant ek_check_script /var/lib/keylime/check_ek_script.sh"
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/test.sh:53]: Please do this configuration update, including the script creation above in the setup phase above. Better to have all required setup at one place.

File: functional/ek-cert-use-ek_check_script/test.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        limeLogfileSubmit $(limeRegistrarLogfile)
+        limeLogfileSubmit $(limeAgentLogfile)
+        if limeTPMEmulated; then
+            rlRun "limeStopIMAEmulator"
+            limeLogfileSubmit $(limeIMAEmulatorLogfile)
+            rlRun "limeStopTPMEmulator"
+            rlServiceRestore tpm2-abrmd
+        fi
+        limeClearData
+        limeRestoreConfig
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/test.sh:74]: Here you should also delete the ek_ceck script created previously.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf [hunk 7]
```diff
@@ -0,0 +1,34 @@
+summary:  Test of keylime checking of endorsement key by manually generate ek and verify against CA certs
+description: |
+    Running all services on localhost.
+    Uses certificates generated by keylime.
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf:4]: I guess this applies to TLS certs, could you please add it (or maybe even remove, it is not important for test scenario).

File: functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    Manually generate endorsement key and add 
+    address of key to keylime.conf.
+    Starts verifier, registrar, agent.
+    Add empty CA certs, expected fail while adding agent.
+    Add right CA certs, pass of adding agent.
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf:9]: Add incorrect CA certs for EK cert validation, expect agent addition to fail.
Add correct CA certs for EK cert validation, expect agent addition to pass.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            rlServiceStart tpm2-abrmd
+            sleep 5
+            # start ima emulator
+            rlRun "limeInstallIMAConfig"
+            rlRun "limeStartIMAEmulator"
+        else
+            rlDie "Test work only when TPM is emulated."
+        fi
+        #change tpm_ownerpassword    
+        rlRun "limeUpdateConf cloud_agent tpm_ownerpassword $PASSWORD_TPM"
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh:35]: Please add this line below line 48. Also, better use comment "Configure TPM owner password", we are not really changing it here.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        rlRun "cat $CACERTDIR/issuercert.pem $CACERTDIR/swtpm-localca-rootca-cert.pem > bundle.pem"
+        rlRun "cp bundle.pem $CERTDIR"
+        rlRun -s "keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u $AGENT_ID --allowlist allowlist.txt --exclude excludelist.txt -f excludelist.txt -c update"
+        rlRun "limeWaitForAgentStatus $AGENT_ID 'Get Quote'" 
+        rlRun -s "keylime_tenant -c cvlist"
+        rlAssertGrep "{'code': 200, 'status': 'Success', 'results': {'uuids':.*'$AGENT_ID'" $rlRun_LOG -E
+    rlPhaseEnd
+
+    rlPhaseStartCleanup "Do the keylime cleanup"
+        rlRun "limeStopAgent"
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh:84]: I am thinking whether we need to add some extra clean up steps. Maybe not.

**Comment by kkaarreell** [comment]
Great work! I have couple of nitpicks. Also, please add this test to a test plan for rust agent [plans/rust-keylime-tests.fmf](https://github.com/RedHat-SP-Security/keylime-tests/blob/main/plans/rust-keylime-tests.fmf)

File: functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+rlJournalStart
+
+    rlPhaseStartSetup "Do the keylime setup"
+        rlRun 'rlImport "./test-helpers"' || rlDie "cannot import keylime-tests/test-helpers library"
+        rlAssertRpm keylime
+        # update /etc/keylime.conf
+        limeBackupConfig       
+        # if TPM emulator is present
+        if rlIsRHEL '<9'; then
```
[the remark this answers is not in the document]
Comment (Koncpa) [reply functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh]: I'm sure, that I have set the right condition in .fmf file, but test still run on RHEL-8. When I comment part with this condition, the test possibly fail on RHEL-8.

File: functional/ek-cert-use-ek_check_script/test.sh (hunk 3 above)
Comment (Koncpa) [reply functional/ek-cert-use-ek_check_script/test.sh]: Useless comment related with previous debugging. Removed.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  - keylime
+  - keylime-verifier
+  - keylime-registrar
+  - python3-keylime-agent
+  - keylime-tenant
+  - keylime-tools
+duration: 5m
+enabled: true
+adjust:
+  - when: distro == rhel-8 or distro = centos-stream-8
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf:30]: Maybe == for c8s?

**Comment by kkaarreell** [comment]
Please remember to update plan for rust agent, we need to ensure this work with it too.

**Comment by Koncpa** [comment]
> Please remember to update plan for rust agent, we need to ensure this work with it too.

Yeah sure, I added also to this plan.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf (hunk 8 above)
Comment (Koncpa) [reply functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf:9]: Thank you for remind me, I update the description in editor, but forgot to save it, so I overlooked it.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf (hunk 7 above)
Comment (Koncpa) [reply functional/ek-cert-use-ek_handle-custom-ca_certs/main.fmf]: Yeah, I removed it, it's not necessary to describe here.

File: functional/ek-cert-use-ek_check_script/test.sh [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        rlRun "limeWaitForVerifier"
+        rlRun "limeStartRegistrar"
+        rlRun "limeWaitForRegistrar"
+        rlRun "limeStartAgent"
+        rlRun "limeWaitForAgentRegistration ${AGENT_ID}"
+        # create allowlist and excludelist
+        limeCreateTestLists
+    rlPhaseEnd
+
+    rlPhaseStartTest "Add keylime agent and check genuine of TPM via ek_check_script option"
```
Comment (kkaarreell) [inline functional/ek-cert-use-ek_check_script/test.sh:53]: @Koncpa Beside the positive scenario we wanted to test a situation when the scripts exits with non-zero exit code. Could you please create a new PR adding this?

**Comment by kkaarreell** [comment]
Thank you for all updates. I am going to merge it, despite the one failure with Rust agent so devs can use this test when working on a fix.

File: functional/ek-cert-use-ek_check_script/test.sh (hunk 13 above)
Comment (Koncpa) [reply functional/ek-cert-use-ek_check_script/test.sh:53]: Yeah, sure @kkaarreell I add PR with failing scenario.

File: functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh (hunk 10 above)
Comment (Koncpa) [reply functional/ek-cert-use-ek_handle-custom-ca_certs/test.sh:81]: Maybe remove authorization values for TPM objects and evict the generated endorsement keys?
